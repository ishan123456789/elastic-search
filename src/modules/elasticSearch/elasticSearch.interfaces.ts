export interface CampaignData {
    campaign_id: string;
    campaign_name: string;
}

export interface ContentData {
    content_id: string;
    content_name: string;
    content_type: ContentType;
}

export enum ContentType {
    VOD = 'VOD',
    LIVE = 'LIVE',
}

export interface AdTagData {
    ad_tag_id: string;
    tag_name: string;
    ad_type: AdType;
    ad_metrics: AdMetrics;
}

export enum AdType {
    VAST = 'VAST',
    VMAP = 'VMAP',
    VAST_PLAYLIST = 'VAST_PLAYLIST',
}

export enum DemandType {
    DIRECT = 'direct',
    PROGRAMMATIC = 'programmatic',
}

export interface AppSchema {
    timestamp: Date;
    appowner_id: string;
    appowner_name: string;
    demand_type: DemandType;
    requests: number;
    campaign_data: CampaignData;
    content_data: ContentData;
    ad_tag_data: AdTagData;
    impressions: number;
    clicks: number;
    conversion: number;
    revenue: number;
}

export interface AdMetrics {
    first_quarter: number;
    mid_point: number;
    third_quarter: number;
    pause: number;
    complete: number;
    skip: number;
}

export interface CampaignReportResponse {
    timestamp: Date | string;
    campaign_id: string;
    campaign_name: string;
    appowner_id: string;
    appowner_name: string;
    demand_type: string;
    requests_count: number;
    impressions_count: number;
    clicks_count: number;
    conversion_count: number;
    revenue_count: number;
    CTR: number;
    eCPM: number;

    ad_tag_id: string;
    tag_name: string;
    ad_type: string;
    content_id: string;
    content_name: string;
    content_type: string;

    first_quarter: number;
    mid_point: number;
    third_quarter: number;
    pause: number;
    complete: number;
    skip: number;
}

export interface ReportQueryOptions {
    dateFrom: Date | string;
    dateTo: Date | string;
    appowner_id: string;
    campaign_id: string;
    demand_type: string;
    ad_tag_id: string;
    ad_type: string;
    content_id: string;
    content_type: string;
    page: number;
    size: number;
    groupBy: string;
    reportType: ReportType;
}
export const APP_INDEX_KEY = 'app';

export enum ReportType {
    CAMPAIGN = 'campaign',
    CONTENT = 'content',
    ADTAG = 'adtag',
}

export enum GroupByKeys {
    timestamp = 'timestamp',
    appowner_id = 'appowner_id',
    ad_tag_id = 'ad_tag_id',
    ad_type = 'ad_type',
    content_id = 'content_id',
    content_type = 'content_type',
    campaign_id = 'campaign_id',
    demand_type = 'demand_type',
}
export const TransformKey = {
    ad_tag_id: 'ad_tag_data.ad_tag_id',
    ad_type: 'ad_tag_data.ad_type',
    tag_name: 'ad_tag_data.tag_name',
    appowner_id: 'appowner_id',
    campaign_id: 'campaign_data.campaign_id',
    campaign_name: 'campaign_data.campaign_name',
    content_id: 'content_data.content_id',
    content_type: 'content_data.content_type',
    content_name: 'content_data.content_name',
    demand_type: 'demand_type',
    appowner_name: 'demand_type',
    timestamp: 'timestamp',
};
export const TranslateToGroupByKey: { [key in GroupByKeys]: string } = {
    ad_tag_id: `doc['${TransformKey.ad_tag_id}']`,
    ad_type: `doc['${TransformKey.ad_type}']`,
    appowner_id: `doc['${TransformKey.appowner_id}']`,
    campaign_id: `doc['${TransformKey.campaign_id}']`,
    content_id: `doc['${TransformKey.content_id}']`,
    content_type: `doc['${TransformKey.content_type}']`,
    demand_type: `doc['${TransformKey.demand_type}']`,
    timestamp: `doc['timestamp'].value.toString('yyyy_MM_dd_hh')`,
};
