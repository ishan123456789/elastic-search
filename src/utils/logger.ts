import pino from 'pino';
import { reqSerializer, resSerializer, errSerializer } from 'koa-req-logger';

const isDev = process.env.NODE_ENV !== 'production';

// https://getpino.io/#/docs/help?id=mapping-pino-log-levels-to-google-cloud-logging-stackdriver-serverity-levels
const PinoLevelToSeverityLookup: { [key: string]: string } = {
    trace: 'DEBUG',
    debug: 'DEBUG',
    info: 'INFO',
    warn: 'WARNING',
    error: 'ERROR',
    fatal: 'CRITICAL',
};

const defaultPinoConf = {
    messageKey: 'message',
    formatters: {
        level(label: string, number: number) {
            return {
                severity: PinoLevelToSeverityLookup[label] || PinoLevelToSeverityLookup.info,
                level: number,
            };
        },
    },
};

export function createLogger(options: pino.LoggerOptions): pino.Logger {
    return pino({ ...options, ...defaultPinoConf });
}

export const logger = createLogger({
    name: 'elastic-search-service',
    level: isDev ? 'debug' : 'info',
    enabled: !(process.env.NO_LOG == 'true'),
});

// THe types aren't updated
export const requestLogger = (logger.child as any)(
    {},
    {
        serializers: {
            req: reqSerializer,
            res: resSerializer,
            err: errSerializer,
        },
    },
);
