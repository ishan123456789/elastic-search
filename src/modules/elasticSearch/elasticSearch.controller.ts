import { Context } from 'koa';
import { logger } from '../../utils/logger';
import { ReportQueryOptions } from './elasticSearch.interfaces';
import { ElasticSearchService } from './elasticSearch.service';
import { reportQueryParamsValidator, validateRouteParam } from './elasticSearch.validator';

export class ElasticSearchController {
    static async getContent(ctx: Context): Promise<void> {
        try {
            const { count } = ctx.query;
            ctx.body = await ElasticSearchService.getAllRecords(+count);
        } catch (err) {
            logger.error({ err }, 'Error while getting records');
            ctx.throw('Error while fetching report');
        }
    }
    static async getReport(ctx: Context): Promise<void> {
        try {
            const queryParams = ctx.query;
            const { reportType } = ctx.params;
            await validateRouteParam.validate(reportType);

            await reportQueryParamsValidator.validate(queryParams, { strict: true });

            ctx.body = await ElasticSearchService.getCampaignReport({
                ...queryParams,
                reportType,
            } as unknown as ReportQueryOptions);
        } catch (err) {
            logger.error({ err }, 'Error while getting records');
            ctx.throw(err.type === 'ValidationError' ? 'Error while fetching report' : err?.message || '');
        }
    }
    static async persistContent(ctx: Context): Promise<void> {
        try {
            const { count: numberOfRecords } = ctx.query;
            const processedAtOnceThreshold = 5000;

            if (!+numberOfRecords || +numberOfRecords < processedAtOnceThreshold) {
                const count = await ElasticSearchService.generateAndPersistRecords(+numberOfRecords);
                logger.info('Total records increased to ' + count);
                ctx.body = { totalRecordsNow: count };
            } else {
                let toInsertRecords = +numberOfRecords;
                ctx.body = `Batch processing has started check logs for more info`;
                (async () => {
                    let batches = Math.ceil(+numberOfRecords / processedAtOnceThreshold);
                    while (toInsertRecords > 1) {
                        const count = await ElasticSearchService.generateAndPersistRecords(
                            Math.min(processedAtOnceThreshold, toInsertRecords),
                            batches,
                        );
                        logger.info(`Count now ${count}`);
                        toInsertRecords = toInsertRecords - processedAtOnceThreshold;
                        logger.info('Pending records ' + toInsertRecords);
                        batches--;
                    }
                })();
            }
        } catch (err) {
            logger.error({ err }, 'Error while create bulk records');
            ctx.throw('Error while bulk persisting');
        }
    }
}
