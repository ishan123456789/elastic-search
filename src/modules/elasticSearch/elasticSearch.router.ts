import { Context, DefaultState } from 'koa';
import Router from 'koa-router';
import { ElasticSearchController } from './elasticSearch.controller';

const router = new Router<DefaultState, Context>();

router.get('/', ElasticSearchController.getContent).get('/persist', ElasticSearchController.persistContent);

export const elasticSearchRoutes = router.routes();
