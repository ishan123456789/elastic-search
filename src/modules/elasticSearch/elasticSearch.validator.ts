import moment from 'moment';
import * as yup from 'yup';
import { AdType, ContentType, DemandType, GroupByKeys, ReportType } from './elasticSearch.interfaces';

export const isValidDate = (value: any): boolean => {
    if (!value) return false;
    return moment(value, true).isValid();
};
export const validateRouteParam = yup
    .string()
    .oneOf(Object.values(ReportType))
    .required()
    .typeError('Invalid route param should be one of campaign || content || adtag');
export const reportQueryParamsValidator = yup
    .object()
    .shape({
        dateFrom: yup.string().test('isValid', 'From date is invalid', isValidDate),
        dateTo: yup.string().test('isValid', 'To date is invalid', isValidDate),

        appowner_id: yup.string().nullable(),
        campaign_id: yup.string().nullable(),
        ad_tag_id: yup.string().nullable(),
        content_id: yup.string().nullable(),
        after_key: yup.string().nullable(),

        ad_type: yup.mixed().oneOf(Object.values(AdType)),
        content_type: yup.mixed().oneOf(Object.values(ContentType)),
        demand_type: yup.mixed().oneOf(Object.values(DemandType)),

        page: yup.string().matches(/^\d+$/),
        size: yup.string().matches(/^\d+$/),
        sort: yup.mixed().oneOf(['asc', 'desc']),
        groupBy: yup.string().test(function (value: any) {
            if (!value) return true;
            return value.split(',').find((key: any) => !(Object.values(GroupByKeys).indexOf(key) === -1));
        }),
    })
    .noUnknown();
