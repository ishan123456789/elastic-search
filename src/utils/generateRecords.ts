import moment from 'moment';
import { AdType, AppSchema, ContentType, DemandType } from '../modules/elasticSearch/elasticSearch.interfaces';

const maxAppOwnerCount = 1;
const maxCampaignCount = 10;
const maxContentIds = 100;
const maxAdTagIds = 1000;

const maxCount = 10 ** 6;

export const randomIntegerToStartIdsWith = 1111;
// Generates a number between 0 to n
const genRandomNumber = (upto: number) => Math.floor(Math.random() * upto || 10);

export const getCurrentId = (count: number, max: number): string =>
    Math.floor((count * max) / maxCount) + max + randomIntegerToStartIdsWith + '';

export const getDemandType = (): DemandType => {
    return genRandomNumber(100) > 90 ? DemandType.PROGRAMMATIC : DemandType.DIRECT;
};

const generateRecord = (count: number): AppSchema => {
    return {
        timestamp: moment().subtract(genRandomNumber(100), 'days').toDate(),
        appowner_id: getCurrentId(count, maxAppOwnerCount),
        demand_type: getDemandType(),
        requests: genRandomNumber(100000) + '',
        campaign_data: {
            campaign_id: getCurrentId(count, maxCampaignCount),
            campaign_name: `campaign ${getCurrentId(count, maxCampaignCount)}`,
        },
        content_data: {
            content_id: getCurrentId(count, maxContentIds),
            content_name: `Web series EP ${getCurrentId(count, maxContentIds)}`,
            content_type: genRandomNumber(100) > 50 ? ContentType.LIVE : ContentType.VOD,
        },
        ad_tag_data: {
            ad_tag_id: getCurrentId(count, maxAdTagIds),
            tag_name: `automobile tag ${getCurrentId(count, maxAdTagIds)}`,
            ad_type: Object.values(AdType)[genRandomNumber(2)],
            ad_metrics: {
                first_quarter: 1300 + genRandomNumber(1500),
                mid_point: 1400 + genRandomNumber(1500),
                third_quarter: 1415 + genRandomNumber(1500),
                pause: 4574 + genRandomNumber(1500),
                complete: 8546 + genRandomNumber(1500),
                skip: 12314 + genRandomNumber(1500),
            },
        },
        impressions: genRandomNumber(10000),
        clicks: genRandomNumber(1000),
        conversion: genRandomNumber(100),
        revenue: genRandomNumber(10000),
    };
};

export const generateRecords = (count: number, N: number): Array<AppSchema> => {
    return Array(N || 1000)
        .fill(null)
        .map((_, i) => generateRecord(count + i));
};
