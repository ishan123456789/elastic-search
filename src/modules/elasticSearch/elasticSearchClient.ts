import { Client } from '@elastic/elasticsearch';
import { logger } from '../../utils/logger';
import { APP_INDEX_KEY } from './elasticSearch.interfaces';
// Export this to external file outside this module if used in different module as well
export const elasticSearchClient = new Client({
    node: process.env.ELASTIC_SEARCH_URL || 'https://localhost:9200',
});

export const bulkInsertHelper = (index: string, doc: Array<{ [key: string]: any }>): Array<any> => {
    return doc.flatMap((doc) => [{ index: { _index: index } }, doc]);
};

(async () => {
    try {
        // Un comment and comment again to have index deleted and created again
        // await elasticSearchClient.indices.delete({
        //     index: APP_INDEX_KEY,
        // });
        // return;
        // await elasticSearchClient.indices.refresh({ index: APP_INDEX_KEY });

        const { body: exists } = await elasticSearchClient.indices.exists({
            index: APP_INDEX_KEY,
        });
        if (exists) {
            logger.info('Index already exists');
            return;
        }
        const result = await elasticSearchClient.indices.create({
            index: APP_INDEX_KEY,
            body: {
                mappings: {
                    properties: {
                        appowner_id: {
                            type: 'text',
                            fielddata: true,
                            fields: {
                                raw: { type: 'keyword' },
                            },
                        },
                        demand_type: { type: 'keyword' },
                        campaign_data: {
                            type: 'nested',
                            include_in_parent: true,
                            properties: {
                                campaign_id: {
                                    type: 'text',
                                    fielddata: true,
                                    fields: {
                                        raw: { type: 'keyword' },
                                    },
                                },
                            },
                        },
                        content_data: {
                            type: 'nested',
                            include_in_parent: true,
                            properties: {
                                content_id: {
                                    type: 'text',
                                    fielddata: true,
                                    fields: {
                                        raw: { type: 'keyword' },
                                    },
                                },
                                content_type: { type: 'keyword' },
                            },
                        },
                        ad_tag_data: {
                            type: 'nested',
                            include_in_parent: true,
                            properties: {
                                ad_tag_id: {
                                    type: 'text',
                                    fielddata: true,
                                    fields: {
                                        raw: { type: 'keyword' },
                                    },
                                },
                                ad_type: { type: 'keyword' },
                            },
                        },
                    },
                },
            },
        });
        logger.info({ result }, 'Index created');
    } catch (err) {
        logger.error({ err }, 'Error while creating index');
    }
})();
