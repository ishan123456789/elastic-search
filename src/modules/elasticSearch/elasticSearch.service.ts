import { transformElasticSearchOutput, transformQueryParams } from './helpers/elasticSearch.utils';
import { generateRecords } from '../../utils/generateRecords';
import { logger } from '../../utils/logger';
import { AppSchema, APP_INDEX_KEY, ReportResponse, ReportQueryOptions, TransformKey } from './elasticSearch.interfaces';
import { bulkInsertHelper, elasticSearchClient } from './elasticSearchClient';

export class ElasticSearchService {
    static async getCampaignReport(options: ReportQueryOptions): Promise<ReportResponse[]> {
        await elasticSearchClient.indices.refresh({ index: APP_INDEX_KEY });
        const { query, from, size, groupBy } = transformQueryParams(options);
        logger.debug({ query, from, size, groupBy }, 'transformation result');
        const { body } = await elasticSearchClient.search({
            index: APP_INDEX_KEY,
            size: 0,
            body: {
                ...(Object.keys(query).length !== 0 && { query }),
                // https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-datehistogram-aggregation.html#date-histogram-scripts
                runtime_mappings: {
                    'timestamp.group_by': {
                        type: 'keyword',
                        // change the aggregation here to combine multiple
                        script: `emit(${groupBy})`,
                    },
                },
                aggs: {
                    grouped_by_app_owner_id: {
                        terms: {
                            field: 'timestamp.group_by',
                            // https://stackoverflow.com/a/59547465/6517383
                            size: (from + 1) * size,
                        },

                        aggs: {
                            // https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-top-hits-aggregation.html
                            top_docs: {
                                top_hits: {
                                    size: 1,
                                    _source: {
                                        include: Object.values(TransformKey),
                                    },
                                },
                            },
                            // TODO: Change the below to sum to see if speed improves
                            clicks_stats: {
                                stats: { field: 'clicks' },
                            },
                            conversion_stats: {
                                stats: { field: 'conversion' },
                            },
                            revenue_stats: {
                                stats: { field: 'revenue' },
                            },
                            impressions_stats: {
                                stats: { field: 'impressions' },
                            },
                            requests_stats: {
                                stats: { field: 'requests' },
                            },

                            first_quarter_stats: {
                                stats: { field: 'ad_tag_data.ad_metrics.first_quarter' },
                            },
                            mid_point_stats: {
                                stats: { field: 'ad_tag_data.ad_metrics.mid_point' },
                            },
                            third_quarter_stats: {
                                stats: { field: 'ad_tag_data.ad_metrics.third_quarter' },
                            },
                            pause_stats: {
                                stats: { field: 'ad_tag_data.ad_metrics.pause' },
                            },
                            complete_stats: {
                                stats: { field: 'ad_tag_data.ad_metrics.complete' },
                            },
                            skip_stats: {
                                stats: { field: 'ad_tag_data.ad_metrics.skip' },
                            },

                            bucket_paginate: {
                                bucket_sort: {
                                    from,
                                    size,
                                },
                            },
                        },
                    },
                },
            },
        });

        return transformElasticSearchOutput(body?.aggregations as any);
    }

    static async getAllRecords(size?: number): Promise<Array<AppSchema>> {
        await elasticSearchClient.indices.refresh({ index: APP_INDEX_KEY });
        const { body } = await elasticSearchClient.search({
            index: APP_INDEX_KEY,
            size: size || 100,
            body: {
                // query: {
                //     match: {},
                // },
            },
        });
        return body?.hits || [];
    }
    static async generateAndPersistRecords(numberOfRecords: number, batches?: number): Promise<number> {
        const startedAt = Date.now();
        const { body: { count: initialCount } = {} } = await elasticSearchClient.count({
            index: APP_INDEX_KEY,
        });
        const records = generateRecords(initialCount, +numberOfRecords);
        logger.debug(`To Insert records count ${numberOfRecords}`);
        // logger.debug(records, 'To Insert records'); // Uncomment when record count is low
        await elasticSearchClient.bulk({
            body: bulkInsertHelper(APP_INDEX_KEY, records),
        });
        await elasticSearchClient.indices.refresh({ index: APP_INDEX_KEY });
        const {
            body: { count },
        } = await elasticSearchClient.count({
            index: APP_INDEX_KEY,
        });
        const timeTaken = Math.floor((Date.now() - startedAt) / 1000);
        logger.debug('Total records increased to ' + count);
        logger.debug(`Took ${timeTaken} seconds`);
        logger.debug((batches && `Expected total time: ${timeTaken * batches}`) || '');
        return count;
    }
}
