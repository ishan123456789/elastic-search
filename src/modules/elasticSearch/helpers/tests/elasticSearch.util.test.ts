import { transformElasticSearchOutput, transformQueryParams } from '../elasticSearch.utils';
import { queryParamsFixture } from './queryPrams.fixture';
import { elasticSearchResponseFixture } from './response.fixture';

describe('Tests transformElasticSearchOutput', () => {
    // Check __snapshots__ folder to see if it's as expected
    it('Matches snapshot', () => {
        const result = transformElasticSearchOutput(elasticSearchResponseFixture as any);
        // Check snapshot to confirm the result is right
        expect(result).toMatchSnapshot();
    });
});

describe('Tests transformQueryParam', () => {
    it('Matches snapshot', () => {
        const result = transformQueryParams(queryParamsFixture as any);
        // Check snapshot to confirm the result is right
        expect(result).toMatchSnapshot();
    });
});
