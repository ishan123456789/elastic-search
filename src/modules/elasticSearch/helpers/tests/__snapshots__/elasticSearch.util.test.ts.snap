// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests transformElasticSearchOutput Matches snapshot 1`] = `
Array [
  Object {
    "CTR": 107.30474198047419,
    "ad_tag_id": "3123",
    "ad_type": "VMAP",
    "appowner_id": "1113",
    "appowner_name": undefined,
    "campaign_id": "1131",
    "campaign_name": "campaign 1131",
    "clicks_count": 4924,
    "complete": 83151,
    "content_id": "1312",
    "content_name": "Web series EP 1312",
    "content_type": "LIVE",
    "conversion_count": 342,
    "demand_type": "direct",
    "eCPM": 0.0007727724356275576,
    "first_quarter": 17911,
    "impressions_count": 45888,
    "mid_point": 17608,
    "pause": 47129,
    "requests_count": 505905,
    "revenue_count": 59381,
    "skip": 120143,
    "tag_name": "automobile tag 3123",
    "third_quarter": 19946,
    "timestamp": "2021-08-03T16:52:02.517Z",
  },
  Object {
    "CTR": 99.82155754769087,
    "ad_tag_id": "2516",
    "ad_type": "VAST",
    "appowner_id": "1112",
    "appowner_name": undefined,
    "campaign_id": "1125",
    "campaign_name": "campaign 1125",
    "clicks_count": 4699,
    "complete": 83556,
    "content_id": "1251",
    "content_name": "Web series EP 1251",
    "content_type": "LIVE",
    "conversion_count": 362,
    "demand_type": "direct",
    "eCPM": 0.001326663472648874,
    "first_quarter": 18560,
    "impressions_count": 47074,
    "mid_point": 18449,
    "pause": 48886,
    "requests_count": 472604,
    "revenue_count": 35483,
    "skip": 118744,
    "tag_name": "automobile tag 2516",
    "third_quarter": 20265,
    "timestamp": "2021-08-04T14:18:56.497Z",
  },
  Object {
    "CTR": 94.04339427698666,
    "ad_tag_id": "3451",
    "ad_type": "VMAP",
    "appowner_id": "1113",
    "appowner_name": undefined,
    "campaign_id": "1134",
    "campaign_name": "campaign 1134",
    "clicks_count": 4187,
    "complete": 85000,
    "content_id": "1345",
    "content_name": "Web series EP 1345",
    "content_type": "VOD",
    "conversion_count": 631,
    "demand_type": "direct",
    "eCPM": 0.0008844434732513558,
    "first_quarter": 19291,
    "impressions_count": 44522,
    "mid_point": 21315,
    "pause": 45452,
    "requests_count": 482440,
    "revenue_count": 50339,
    "skip": 116190,
    "tag_name": "automobile tag 3451",
    "third_quarter": 21199,
    "timestamp": "2021-08-04T16:56:20.717Z",
  },
  Object {
    "CTR": 80.10168708111856,
    "ad_tag_id": "3580",
    "ad_type": "VMAP",
    "appowner_id": "1113",
    "appowner_name": undefined,
    "campaign_id": "1135",
    "campaign_name": "campaign 1135",
    "clicks_count": 3466,
    "complete": 82280,
    "content_id": "1357",
    "content_name": "Web series EP 1357",
    "content_type": "VOD",
    "conversion_count": 430,
    "demand_type": "direct",
    "eCPM": 0.0010687909102136595,
    "first_quarter": 20031,
    "impressions_count": 43270,
    "mid_point": 19775,
    "pause": 46051,
    "requests_count": 455191,
    "revenue_count": 40485,
    "skip": 116075,
    "tag_name": "automobile tag 3580",
    "third_quarter": 17982,
    "timestamp": "2021-08-04T16:57:45.413Z",
  },
  Object {
    "CTR": 124.18554476806904,
    "ad_tag_id": "3584",
    "ad_type": "VAST",
    "appowner_id": "1113",
    "appowner_name": undefined,
    "campaign_id": "1135",
    "campaign_name": "campaign 1135",
    "clicks_count": 5756,
    "complete": 82417,
    "content_id": "1358",
    "content_name": "Web series EP 1358",
    "content_type": "LIVE",
    "conversion_count": 474,
    "demand_type": "direct",
    "eCPM": 0.0014648715274485635,
    "first_quarter": 18617,
    "impressions_count": 46350,
    "mid_point": 20891,
    "pause": 48616,
    "requests_count": 452267,
    "revenue_count": 31641,
    "skip": 119426,
    "tag_name": "automobile tag 3584",
    "third_quarter": 20752,
    "timestamp": "2021-08-04T16:57:48.778Z",
  },
  Object {
    "CTR": 57.92351965262274,
    "ad_tag_id": "2126",
    "ad_type": "VMAP",
    "appowner_id": "1112",
    "appowner_name": undefined,
    "campaign_id": "1121",
    "campaign_name": "campaign 1121",
    "clicks_count": 2828,
    "complete": 75863,
    "content_id": "1212",
    "content_name": "Web series EP 1212",
    "content_type": "VOD",
    "conversion_count": 338,
    "demand_type": "direct",
    "eCPM": 0.000994500234249282,
    "first_quarter": 15809,
    "impressions_count": 48823,
    "mid_point": 16585,
    "pause": 42707,
    "requests_count": 519210,
    "revenue_count": 49093,
    "skip": 103898,
    "tag_name": "automobile tag 2126",
    "third_quarter": 14984,
    "timestamp": "2021-07-28T14:12:58.153Z",
  },
  Object {
    "CTR": 93.55576739752145,
    "ad_tag_id": "2129",
    "ad_type": "VMAP",
    "appowner_id": "1112",
    "appowner_name": undefined,
    "campaign_id": "1121",
    "campaign_name": "campaign 1121",
    "clicks_count": 4907,
    "complete": 74380,
    "content_id": "1212",
    "content_name": "Web series EP 1212",
    "content_type": "LIVE",
    "conversion_count": 554,
    "demand_type": "direct",
    "eCPM": 0.001125077757995667,
    "first_quarter": 15124,
    "impressions_count": 52450,
    "mid_point": 17099,
    "pause": 43484,
    "requests_count": 466548,
    "revenue_count": 46619,
    "skip": 105501,
    "tag_name": "automobile tag 2129",
    "third_quarter": 19178,
    "timestamp": "2021-07-28T14:12:58.178Z",
  },
  Object {
    "CTR": 93.73236558930552,
    "ad_tag_id": "2179",
    "ad_type": "VAST",
    "appowner_id": "1112",
    "appowner_name": undefined,
    "campaign_id": "1121",
    "campaign_name": "campaign 1121",
    "clicks_count": 4817,
    "complete": 76066,
    "content_id": "1217",
    "content_name": "Web series EP 1217",
    "content_type": "VOD",
    "conversion_count": 393,
    "demand_type": "direct",
    "eCPM": 0.0016132282772476142,
    "first_quarter": 16342,
    "impressions_count": 51391,
    "mid_point": 17027,
    "pause": 42271,
    "requests_count": 437540,
    "revenue_count": 31856,
    "skip": 104413,
    "tag_name": "automobile tag 2179",
    "third_quarter": 19102,
    "timestamp": "2021-07-28T14:13:47.798Z",
  },
  Object {
    "CTR": 81.4407522511754,
    "ad_tag_id": "2357",
    "ad_type": "VAST",
    "appowner_id": "1112",
    "appowner_name": undefined,
    "campaign_id": "1123",
    "campaign_name": "campaign 1123",
    "clicks_count": 3066,
    "complete": 74246,
    "content_id": "1235",
    "content_name": "Web series EP 1235",
    "content_type": "LIVE",
    "conversion_count": 414,
    "demand_type": "direct",
    "eCPM": 0.0009112186857073702,
    "first_quarter": 15097,
    "impressions_count": 37647,
    "mid_point": 17987,
    "pause": 44237,
    "requests_count": 416379,
    "revenue_count": 41315,
    "skip": 105567,
    "tag_name": "automobile tag 2357",
    "third_quarter": 15717,
    "timestamp": "2021-07-28T14:16:32.371Z",
  },
  Object {
    "CTR": 106.29179252833501,
    "ad_tag_id": "2393",
    "ad_type": "VAST",
    "appowner_id": "1112",
    "appowner_name": undefined,
    "campaign_id": "1123",
    "campaign_name": "campaign 1123",
    "clicks_count": 4117,
    "complete": 75109,
    "content_id": "1239",
    "content_name": "Web series EP 1239",
    "content_type": "VOD",
    "conversion_count": 373,
    "demand_type": "direct",
    "eCPM": 0.0011045427324835315,
    "first_quarter": 15005,
    "impressions_count": 38733,
    "mid_point": 18169,
    "pause": 42775,
    "requests_count": 384092,
    "revenue_count": 35067,
    "skip": 104625,
    "tag_name": "automobile tag 2393",
    "third_quarter": 16759,
    "timestamp": "2021-07-28T14:17:03.560Z",
  },
]
`;

exports[`Tests transformQueryParam Matches snapshot 1`] = `
Object {
  "from": 30,
  "groupBy": "doc['timestamp'].value.toString('yyyy_MM_dd_hh')+doc['ad_tag_data.ad_tag_id']+doc['ad_tag_data.ad_type']+doc['appowner_id']+doc['campaign_data.campaign_id']+doc['content_data.content_id']+doc['content_data.content_type']+doc['demand_type']",
  "query": Object {
    "bool": Object {
      "filter": Array [
        Object {
          "range": Object {
            "timestamp": Object {
              "gt": 2021-07-27T18:30:00.000Z,
              "lt": 2021-08-29T18:30:00.000Z,
            },
          },
        },
        Object {
          "multi_match": Object {
            "fields": Array [
              "ad_tag_data.ad_tag_id",
              "campaign_data.campaign_id",
            ],
            "query": "123 321",
          },
        },
      ],
    },
  },
  "size": 10,
}
`;
